rotkehlchen/chain/evm/decoding/pendle/decoder.py

```diff
@@ -1,122 +1,288 @@
-                token_decimals=DEFAULT_TOKEN_DECIMALS,
-            )
-            for event in context.decoded_events:
-                if (
-                    event.amount == amount and
-                    event.address == self.ve_pendle_contract and
-                    event.event_type == HistoryEventType.RECEIVE and
-                    event.event_subtype == HistoryEventSubType.NONE
-                ):
+                    event.event_type = HistoryEventType.RECEIVE
+                    event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                    event.notes = f'Receive {amount_1} {token_1.symbol} from depositing into Pendle'  # noqa: E501
+                elif event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE:  # noqa: E501
                     event.counterparty = CPT_PENDLE
-                    event.event_type = HistoryEventType.WITHDRAWAL
-                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
-                    event.notes = f'Withdraw {amount} PENDLE from vote escrow'
-                    break
-            else:
-                log.error(f'Could not find pendle unlock transfer for transaction {context.transaction}')  # noqa: E501
+                    event.event_type = HistoryEventType.SPEND
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.notes = f'Return {event.amount} {token_1.symbol} to Pendle'
 
         return DEFAULT_DECODING_OUTPUT
 
+    def _decode_pendle_swap(self, context: DecoderContext) -> DecodingOutput:
+        """This decodes swap events that use the PendleSwap contract.
+
+        Below are aggregators used by PendleSwap with their indexes:
+        Kyberswap -> 1
+        Odos -> 2
+        Okx -> 4
+        1inch -> 5
+
+        Data is gotten from:
+        https://github.com/pendle-finance/pendle-core-v2-public/blob/761033bcc36ecfdb3c523458d7f66aeabb01c3ab/contracts/router/swap-aggregator/IPSwapAggregator.sol#L18
+        """
+        if context.tx_log.topics[0] != SWAP_SINGLE_TOPIC:
+            return DEFAULT_DECODING_OUTPUT
+
+        # Determine swap type and retrieve received token details.
+        # Pendle only emits an event for the token being swapped,
+        # so we must check transaction logs to find the received token
+        amount_out = asset_normalized_value(
+            amount=int.from_bytes(context.tx_log.data[:32]),
+            asset=(token_out := self.base.get_token_or_native(bytes_to_address(context.tx_log.topics[2]))),  # noqa: E501
+        )
+        if (swap_type := int.from_bytes(context.tx_log.topics[1])) == 1:
+            expected_log_topic = KYBERSWAP_SWAPPED_TOPIC
+        elif swap_type == 2:
+            expected_log_topic = ODOS_SWAP_TOPIC
+        elif swap_type == 4:
+            expected_log_topic = OKX_ORDER_RECORD_TOPIC
+        else:  # no transaction was found that touches 1inch so we don't check for it
```

Comment:
What do you mean here? Can you make a note and a TODO for this? I guess you could solve it in a subsequent PR. Perhaps by asking in Pendle's chat/discord?

Reply:
asked in the pendle discord. created a note for it and added it to the todo comment.